challenge-manager/chain-watcher/watcher.go

```diff
@@ -1046,19 +1053,90 @@ func (w *Watcher) processEdgeConfirmation(
 	// Check if we should confirm the assertion by challenge winner.
 	challengeLevel := edge.GetChallengeLevel()
 	if challengeLevel == protocol.NewBlockChallengeLevel() {
-		if confirmAssertionErr := w.chain.ConfirmAssertionByChallengeWinner(ctx, protocol.AssertionHash{Hash: common.Hash(claimId)}, edgeId); confirmAssertionErr != nil {
-			return confirmAssertionErr
-		}
-		srvlog.Info("Assertion confirmed by challenge win", log.Ctx{
-			"challengeParentAssertionHash": containers.Trunc(challengeParentAssertionHash.Bytes()),
-		})
+		return w.confirmAssertionByChallengeWinner(ctx, edge, claimId, challengeParentAssertionHash)
 	}
 
 	chal.confirmedLevelZeroEdgeClaimIds.Put(claimId, edge.Id())
 	w.challenges.Put(challengeParentAssertionHash, chal)
 	return nil
 }
 
+func (w *Watcher) confirmAssertionByChallengeWinner(ctx context.Context, edge protocol.SpecEdge, claimId protocol.ClaimId, challengeParentAssertionHash protocol.AssertionHash) error {
+	edgeConfirmedAtBlock, err := edge.ConfirmedAtBlock(ctx)
+	if err != nil {
+		return err
+	}
+	challengeGracePeriodBlocks, err := w.chain.RollupUserLogic().RollupUserLogicCaller.ChallengeGracePeriodBlocks(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return fmt.Errorf("could not get challenge grace period blocks: %w", err)
+	}
+	for {
+		ticker := time.NewTicker(w.assertionConfirmingInterval)
+		defer ticker.Stop()
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			confrimed, err := w.tryConfirmingChallengeByWinner(
+				ctx,
+				edgeConfirmedAtBlock,
+				challengeGracePeriodBlocks,
+				claimId,
+				challengeParentAssertionHash,
+				edge,
+			)
+			if err != nil {
+				return err
+			}
+			if confrimed {
+				return nil
+			}
+		}
+	}
+}
+
+func (w *Watcher) tryConfirmingChallengeByWinner(
+	ctx context.Context,
+	edgeConfirmedAtBlock uint64,
+	challengeGracePeriodBlocks uint64,
+	claimId protocol.ClaimId,
+	challengeParentAssertionHash protocol.AssertionHash,
+	edge protocol.SpecEdge,
+) (bool, error) {
+	latestHeader, err := w.chain.Backend().HeaderByNumber(ctx, nil)
```

Comment:
Would recommend we calculate this inside of confirmAssertionByChallengeWinner, before we enter the for loop. We can then pass it in as an argument to tryConfirmingChallengeWinner. When doing so, we should use the retry package so we make sure it eventually succeeds in case the provider disconnects

Reply:
Note that this function won't work if the parent assertion is not the latest confirmed. See the conditions in #545. We must also check that creationInfo.ParentAssertionHash == LatestConfirmed(). The function looks very similar to the on in the scanner.go file from #545. Maybe we can deduplicate it somehow? Perhaps it can take in a winner assertion hash as an option.Option[protocol.AssertionHash] and if it IsSome(), we check for the grace period and try to confirm by challenge win instead of by time

Reply:
Done.